src/main/scala/mesosphere/marathon/core/storage/impl/zk/ZkPersistenceStore.scala

```diff
@@ -0,0 +1,170 @@
+package mesosphere.marathon.core.storage.impl.zk
+
+import java.time.OffsetDateTime
+
+import akka.actor.Scheduler
+import akka.http.scaladsl.marshalling.Marshaller
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import akka.{ Done, NotUsed }
+import com.google.protobuf.MessageLite
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.StoreCommandFailedException
+import mesosphere.marathon.core.storage.impl.BasePersistenceStore
+import mesosphere.marathon.core.storage.{ CategorizedKey, MarathonProto, MarathonState }
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.util.{ Retry, toRichFuture }
+import mesosphere.util.state.zk.{ Children, GetData, RichCuratorFramework }
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.NoNodeException
+import org.apache.zookeeper.data.Stat
+
+import scala.async.Async.{ async, await }
+import scala.collection.immutable.Seq
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success }
+
+case class ZkId(category: String, id: String, version: Option[OffsetDateTime]) {
+  private val bucket = id.hashCode % 16
```

Comment:
simply beautiful.   I don't see a way to reverse this for debugging / support.  It appears we save the full ID in the bucket.  I just checked and Exhibitor doesn't provide a way of searching.   Any thoughts here?